src/websockethandler.cpp

```diff
@@ -0,0 +1,178 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "websockethandler.h"
+#include "networkrequest.h"
+#include "settingsholder.h"
+#include "urlopener.h"
+#include "mozillavpn.h"
+
+#include <QtCore/QDebug>
+#include <QAbstractSocket>
+#include <QTimer>
+
+constexpr const char* WEBSOCKET_SERVER_URL = "ws://localhost:3000";
+constexpr const int WEBSOCKET_PING_INTERVAL = 30 * 1000;   // 30s
+constexpr const int WEBSOCKET_RETRY_INTERVAL = 30 * 1000;  // 30s
+
+WebSocketHandler::WebSocketHandler() {
+  connect(&m_webSocket, &QWebSocket::connected, this,
+          &WebSocketHandler::onConnected);
+  connect(&m_webSocket, &QWebSocket::disconnected, this,
+          &WebSocketHandler::onClose);
+  connect(&m_webSocket,
+          QOverload<QAbstractSocket::SocketError>::of(&QWebSocket::error), this,
+          &WebSocketHandler::onError);
+  connect(&m_webSocket, &QWebSocket::pong, this, &WebSocketHandler::onPong);
+}
+
+/**
+ * @brief Opens the websocket connection.
+ *
+ * No-op in case the connection is already open.
+ */
+void WebSocketHandler::open() {
+  if (m_webSocket.state() != QAbstractSocket::UnconnectedState &&
+      m_webSocket.state() != QAbstractSocket::ClosingState) {
+    qDebug()
+        << "Attempted to open a WebSocket connection, but it's already open.";
+    return;
+  }
+
+  qDebug() << "Attemtping to open WebSocket connection";
+  m_reconnectOnClose = true;
+
+  QNetworkRequest request;
+  request.setRawHeader("Authorization",
+                       SettingsHolder::instance()->token().toLocal8Bit());
+  request.setUrl(QUrl(WEBSOCKET_SERVER_URL));
+  m_webSocket.open(request);
+}
+
+/**
+ * @brief Closes the websocket connection.
+ *
+ * No-op in case the connection is already closed.
+ *
+ * Closing the websocket connection through this function does keeps it closed,
+ * unless the connection is explicitly reopened through
+ * `WebSocketHandler::open`.
+ */
+void WebSocketHandler::close() {
+  if (m_webSocket.state() == QAbstractSocket::UnconnectedState ||
+      m_webSocket.state() == QAbstractSocket::ClosingState) {
+    qDebug() << "Attempted to close a WebSocket, but it's already closed.";
+    return;
+  }
+
+  qDebug() << "Closing WebSocket";
+  m_reconnectOnClose = false;
+
+  m_webSocket.close();
+}
+
+/**
+ * @brief Sends a ping to the WebSocket server.
+ *
+ * WebSocket connections may silently be broken. Periodically pinging the server
+ * keeps track of the connection. If verified that the connection has been
+ * broken the connection is properly closed, which triggers new connection
+ * attempts.
+ */
+void WebSocketHandler::ping() {
+  if (!this->m_isAlive) {
+    qDebug() << "WebSocket was disconnected silently. Closing.";
+    m_webSocket.close();
+    return;
+  }
+
+  m_isAlive = false;
```

Comment:
Do we want to add a timeout between a ping and a pong? Something like, if we have not received a pong() in  < 20 seconds, let's consider the connection as gone.

Reply:
That is how this is working. The timeout is 30s. We send a ping and change `m_isAlive` to `false` i.e. we assume the connection is broken. If it is not broken we will get a pong and change `m_isAlive` to `true`. If we get to the next ping without ever getting a pong the connections is assumed to be broken and the socked is closed. That trigger the `onClose` handler which attempts a reconnections.

Reply:
We spoke offline, yes, this makes sense.